### PR TITLE
Add apply set-last-applied subcommand 

### PR DESCRIPTION
**Pull request by shiywang**
implement part of https://github.com/kubernetes/community/pull/287, will rebase after https://github.com/kubernetes/kubernetes/pull/41699 got merged, EDIT: since bug output format has been confirmed, will update the behavior of output format soon  
cc @kubernetes/sig-cli-pr-reviews @AdoHe @pwittrock 

```release-note
Support kubectl apply set-last-applied command to update the applied-applied-configuration annotation
```

### CODE REVIEW

**Comment by shiywang** [comment]
why cc @kubernetes/sig-cli-pr-reviews  doesn't assign to right people ?

**Comment by pwittrock** [comment]
cc'ing doesn't set the reviewers, it just makes sure the right people get notified.

File: pkg/kubectl/cmd/apply_set_last_applied.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/kubernetes/pkg/api/annotations"
+	"k8s.io/kubernetes/pkg/kubectl"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+)
+
+type SetLastAppliedOptions struct {
```
Comment (AdoHe) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: this code snippet duplicate with `view-last-applied`, can refactor if that get merged.

File: pkg/kubectl/cmd/apply_set_last_applied.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+func NewCmdApplySetLastApplied(f cmdutil.Factory, out, err io.Writer) *cobra.Command {
+	options := &SetLastAppliedOptions{Out: out, ErrOut: err}
+	cmd := &cobra.Command{
+		Use:     "set-last-applied -f FILENAME",
+		Short:   "Set latest last-applied-configuration annotations of by file",
+		Long:    applySetLastAppliedLong,
+		Example: applySetLastAppliedExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(options.RunApplySetLastApplied(f, args, cmd, out, err, options))
```
Comment (AdoHe) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: I would like this also follow `Complete` `Validate` and `Run` convention.

File: pkg/kubectl/cmd/apply_set_last_applied.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+	cmdutil.AddDryRunFlag(cmd)
+	cmdutil.AddRecordFlag(cmd)
+	cmdutil.AddPrinterFlags(cmd)
+	usage := "that contains the last-applied-configuration annotations"
+	kubectl.AddJsonFilenameFlag(cmd, &options.FilenameOptions.Filenames, "Filename, directory, or URL to files "+usage)
+
+	return cmd
+}
+
+func (o *SetLastAppliedOptions) RunApplySetLastApplied(f cmdutil.Factory, args []string, cmd *cobra.Command, out, errOut io.Writer, options *SetLastAppliedOptions) error {
```
Comment (AdoHe) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: pls avoid direct cobra dependency.

File: pkg/kubectl/cmd/apply_set_last_applied.go (hunk 1 above)
Comment (shiywang) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: I'll refactor it when all sub-commands get merged.

File: pkg/kubectl/cmd/apply_set_last_applied.go (hunk 2 above)
Comment (shiywang) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: just forget that convention again, yuck....

File: pkg/kubectl/cmd/apply_set_last_applied.go (hunk 3 above)
Comment (shiywang) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: but this `PrintResourceInfoForCommand` function depend on cobra, how can we move out this fucntion or refactor this ?

File: pkg/kubectl/cmd/apply_set_last_applied.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	ErrOut                       io.Writer
+}
+
+var (
+	applySetLastAppliedLong = templates.LongDesc(`
+		Set the latest last-applied-configuration annotations by file.`)
+
+	applySetLastAppliedExample = templates.Examples(`
+		# Set the current file's content to the latest-applied-configuration of this resource.
+		kubectl apply set-last-applied -f deploy.yaml`)
```
Comment (AdoHe) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: which resource to be updated? identified by deploy.yaml?

File: pkg/kubectl/cmd/apply_set_last_applied.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+		annotationsMap := map[string]string{}
+
+		localFile, err := runtime.Encode(codec, info.VersionedObject)
+		if err != nil {
+			return err
+		}
+		annotationsMap[annotations.LastAppliedConfigAnnotation] = string(localFile)
+		metadataMap["annotations"] = annotationsMap
+		objMap["metadata"] = metadataMap
+		jsonString, err := json.Marshal(objMap)
```
Comment (AdoHe) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: I am afraid this is not correct here, we should calculate the final patch between the current object and object with annotation updated.

File: pkg/kubectl/cmd/apply_set_last_applied.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+		if err := info.Get(); err != nil {
+			if errors.IsNotFound(err) {
+				return err
+			} else {
+				return cmdutil.AddSourceToErr(fmt.Sprintf("retrieving current configuration of:\n%v\nfrom server for:", info), info.Source, err)
+			}
+		}
+
+		if !dryRun {
```
Comment (AdoHe) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: should add `dry-run` handling.

File: pkg/kubectl/cmd/apply_set_last_applied.go (hunk 5 above)
Comment (shiywang) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: so, it's like run `apply -f ` first, get the result object then calculate the final patch between current object and the result object I got before ?  currently I think I just set the file's object to annotations then do a json-merge-patch

File: pkg/kubectl/cmd/apply_test.go [hunk 7]
```diff
@@ -59,9 +59,10 @@ func validateApplyArgs(cmd *cobra.Command, args []string) error {
 }
 
 const (
-	filenameRC    = "../../../test/fixtures/pkg/kubectl/cmd/apply/rc.yaml"
-	filenameSVC   = "../../../test/fixtures/pkg/kubectl/cmd/apply/service.yaml"
-	filenameRCSVC = "../../../test/fixtures/pkg/kubectl/cmd/apply/rc-service.yaml"
+	filenameRC        = "../../../test/fixtures/pkg/kubectl/cmd/apply/rc.yaml"
```
Comment (pwittrock) [inline pkg/kubectl/cmd/apply_test.go]: also add an e2e test that does

```sh
kubectl apply -f deployment1.yaml // contains replicas = 2
kubectl apply view-last-applied -f deploymen1.yaml // check the last-applied matches expectations
kubectl apply set-last-applied -f deployment2.yaml // doesn't have replicas
kubectl apply view-last-applied -f deploymen1.yaml // check last-applied has been updated
kubectl scale deployment ... // set replicas = 3
kubectl apply -f deployment3.yaml // doesn't have replicas, and changes the images
kubectl get -f deployment3.yaml // verify replicas still = 3 and image has been updated
```

File: pkg/kubectl/cmd/apply_set_last_applied.go (hunk 5 above)
Comment (AdoHe) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: generally this does the same thing as `kubectl annotate`, i would think the way `annotate` use is more clean.

File: pkg/kubectl/cmd/apply_set_last_applied.go (hunk 6 above)
Comment (AdoHe) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: `dry-run` means just print the object that would be sent~yep that would make sense print out the patch when dry-run

**Comment by AdoHe** [comment]
@pwittrock I am wondering whether we should allow set `last-applied` on objects which are not managed by apply. wdyt?

File: pkg/kubectl/cmd/apply_test.go [hunk 8]
```diff
@@ -180,6 +181,30 @@ func validatePatchApplication(t *testing.T, req *http.Request) {
 	}
 }
 
+func checkPatchString(t *testing.T, req *http.Request) {
+	//TODO: remove hardcoded here
+	checkString := "{\"apiVersion\":\"v1\",\"kind\":\"ReplicationController\",\"metadata\":{\"labels\":{\"name\":\"test-rc\"},\"name\":\"test-rc\",\"namespace\":\"test\"},\"spec\":{\"replicas\":1,\"template\":{\"metadata\":{\"labels\":{\"name\":\"test-rc\"}},\"spec\":{\"containers\":[{\"image\":\"nginx\",\"name\":\"test-rc\",\"ports\":[{\"containerPort\":80}]}]}}}}\n"
```
Comment (shiywang) [inline pkg/kubectl/cmd/apply_test.go]: will update that when I adding e2e tests

**Comment by shiywang** [comment]
@AdoHe @pwittrock updated, ptal, will add e2e test when you think this implementation is ok

**Comment by pwittrock** [comment]
> @pwittrock I am wondering whether we should allow set last-applied on objects which are not managed by apply. wdyt?

Good thought.  Maybe disallow by default, but allow a flag to enable it?  We allow other non-apply commands to se the last-applied-configuration annotation using the --save-config flag.  I don't think that name really makes sense here.  How about `--create-annotation`?

@shiywang Would you add a `--create-annotation` flag, whose default value is false.  Then in the command logic, if the flag is false AND the live object does not have an annotation, fail the command with an error message and non-0 exit code.  The error message should tell the user that they need to specify `--create-annotation` to set the annotation for the first time.  Also please update the help message and example to include this usage.

File: pkg/kubectl/cmd/apply_set_last_applied.go (hunk 5 above)
Comment (pwittrock) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: @AdoHe Hm.  I am not sure I totally agree.  WDYT of of leaving it as is for now, so this makes feature freeze (tomorrow), and we can follow up if we decide we want to change it.

File: pkg/kubectl/cmd/apply_set_last_applied.go [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+		if oringalBuf == nil {
+			return cmdutil.AddSourceToErr(fmt.Sprintf("no last-applied-configuration annotation found on resource: %s\n", info.Name), info.Source, err)
+		}
+
+		//only add to PatchBufferList when changed
+		if !bytes.Equal(stripComments(oringalBuf), stripComments(diffBuf)) {
+			o.PatchBufferList = append(o.PatchBufferList, patchBuf)
+			o.InfoList = append(o.InfoList, info)
+		} else {
+			fmt.Fprintf(o.Out, "set-last-applied %s: no valid changes saved, won't be patch\n", info.Name)
```
Comment (pwittrock) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: Awkward:

`set-last-applied %s: no changes required.\n`

File: pkg/kubectl/cmd/apply_test.go [hunk 10]
```diff
@@ -193,6 +218,78 @@ func walkMapPath(t *testing.T, start map[string]interface{}, path []string) map[
...
+		name, nameRC, pathRC, filePath, expectedErr, expectedOut string
+	}{
+		{
+			name:        "set with exist object",
+			filePath:    filenameRC,
+			expectedErr: "",
+			expectedOut: "replicationcontroller/test-rc\n",
+		},
+		{
+			name:        "set with no-exist object",
```
Comment (pwittrock) [inline pkg/kubectl/cmd/apply_test.go]: Add test for the annotation does not exist on the live object.

File: pkg/kubectl/cmd/apply_test.go [hunk 11]
```diff
@@ -193,6 +218,78 @@ func walkMapPath(t *testing.T, start map[string]interface{}, path []string) map[
...
+			filePath:    filenameNoExistRC,
+			expectedErr: "Error from server (NotFound): the server could not find the requested resource (get replicationcontrollers no-exist)",
```
Comment (pwittrock) [inline pkg/kubectl/cmd/apply_test.go]: Add test for a json file

Comment (pwittrock) [inline pkg/kubectl/cmd/apply_test.go]: Add test for a directory of files

**Comment by pwittrock** [comment]
@shiywang Generally looks good.  Please fix the help and error messages with my comments and I will lgtm.  Since there are already some tests written, we can do a follow up PR with the additional tests I asked for and the e2e tests.  Normally I would have you write the remaining tests in the same PR, but I would like to get this in before feature freeze which is tomorrow.

**Comment by pwittrock** [comment]
@k8s-bot test this

**Comment by pwittrock** [comment]
@shiywang Please rebase to resolve conflicts.

**Comment by shiywang** [comment]
@pwittrock ok, great, will do that

File: pkg/kubectl/cmd/apply_set_last_applied.go [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+		without updating any other parts of the object.`)
+
+	applySetLastAppliedExample = templates.Examples(`
+		# Set the last-applied-configuration of a resource to match the contents of a file.
+		kubectl apply set-last-applied -f deploy.yaml
+
+		# Set the last-applied-configuration of resourcesto match the contents of files in a directory.
+		kubectl apply set-last-applied -f path/
+
+		# Set the last-applied-configuration of a resource to match the contents of a file, will create if resource doesn't have one.
```
Comment (pwittrock) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: `will create the annotation if it does not already exist`

File: pkg/kubectl/cmd/apply_set_last_applied.go [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+			} else {
+				return cmdutil.AddSourceToErr(fmt.Sprintf("retrieving current configuration of:\n%v\nfrom server for:", info), info.Source, err)
+			}
+		}
+		oringalBuf, err := kubectl.GetOriginalConfiguration(info.Mapping, info.Object)
+		if err != nil {
+			return cmdutil.AddSourceToErr(fmt.Sprintf("retrieving current configuration of:\n%v\nfrom server for:", info), info.Source, err)
+		}
+		if oringalBuf == nil && !o.CreateAnnotation {
+			return cmdutil.UsageError(cmd, "no last-applied-configuration annotation found on resource: %s, please use '--create-annotation' to create one", info.Name)
```
Comment (pwittrock) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: > please use '--create-annotation' to create one

How about

> to create the annotation, run the command with --create-annotation

**Comment by shiywang** [comment]
@pwittrock

**Comment by pwittrock** [comment]
@k8s-bot test this

**Comment by shiywang** [comment]
@k8s-bot test this

**Comment by shiywang** [comment]
@pwittrock hi, just can not reach you on slack, sorry, just forget update bazel, could you please let bot tests again ?

**Comment by ymqytw** [comment]
@k8s-bot ok to test

**Comment by shiywang** [comment]
@pwittrock docs genreated, ptal

**Comment by shiywang** [comment]
@k8s-bot unit test this

**Pull request merged**